### PR TITLE
WC-2683 Provide documentation for asset upload

### PR DESCRIPTION
**Pull request by WillTaylorDev**
### Summary

Documentation to include asset direct upload instructions for customers who want to interact with Workers APIs programmatically rather than through wrangler.

### Documentation checklist

- [X] The [documentation style guide](https://developers.cloudflare.com/style-guide/) has been adhered to.
- [X] If a larger change - such as adding a new page- an issue has been opened in relation to any incorrect or out of date information that this PR fixes.

### CODE REVIEW

File: src/content/docs/workers/static-assets/direct-upload.mdx [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+## `Upload manifest`
+
+The asset manifest is a ledger which keeps track of files we want to use in our Worker. This manifest is used to track assets associated with each Worker version, and eliminate the need to upload unchanged files prior to a new upload.
+
+The manifest upload request describes each file which we intend to upload. Each file is its own key representing the file path and name, and is an object which contains metadata about the file.
+
+`hash` represents the first 32 characters of the file’s hash, while `size` is the size (in bytes) of the file.
+
+```bash title="Example manifest upload request"
+curl -X POST https://api.cloudflare.com/client/v4/accounts/:accountId/workers/scripts/:scriptName/assets-upload-session \
```
Comment (irvinebroque) [inline src/content/docs/workers/static-assets/direct-upload.mdx]: This API isn't in our API docs?

Are we adding it?

Comment (GregBrimble) [reply src/content/docs/workers/static-assets/direct-upload.mdx]: ~Already merged — it'll be going out on the next release: https://cflare.co/asset-upload-api~

Oh, sorry, never mind, we did actually forget that one. I'll follow up.

Comment (GregBrimble) [reply src/content/docs/workers/static-assets/direct-upload.mdx]: Take 2: https://cflare.co/assets-upload-api-followup

File: src/content/docs/workers/static-assets/direct-upload.mdx [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+```
+
+### `Limitations`
+
+- Each file must be under 25 MiB
+- The overall manifest must not contain more than 20,000 file entries
+
+## `Upload Static Assets`
+
+Users can upload assets by POSTing to https://api.cloudflare.com/client/v4/accounts/:accountID/workers/assets/upload?base64=true. This endpoint requires files be uploaded using multipart/form data. The contents of each file must be base64 encoded, and the `base64` query param must be set to `true`.
```
Comment (WillTaylorDev) [inline src/content/docs/workers/static-assets/direct-upload.mdx]: Will replace this link with a hyperlink to the api docs once the open api schema for this is merged (https://bitbucket.cfdata.org/projects/EW/repos/edgeworker-config-service/pull-requests/6133/overview)

File: src/content/docs/workers/static-assets/direct-upload.mdx [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+	FileTree,
+	InlineBadge,
+	Render,
+	TabItem,
+	Tabs,
+} from "~/components";
+
+:::note
+
+Directly uploading assets via Worker APIs is an advanced approach that most users won't need. Instead, users are encouraged to interact with these APIs through [Wrangler](/workers/static-assets/get-started/#1-create-a-new-worker-project-using-the-cli).
```
Comment (Oxyjun) [inline src/content/docs/workers/static-assets/direct-upload.mdx]:
```suggestion
Directly uploading assets via Worker APIs is an advanced approach that most users will not need. Instead, we encourage users to interact with these APIs through [Wrangler](/workers/static-assets/get-started/#1-create-a-new-worker-project-using-the-cli).
```

File: src/content/docs/workers/static-assets/direct-upload.mdx [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+```mermaid
+sequenceDiagram
+    participant User
+    participant Workers API
+    User<<->>Workers API: Submit manifest<br/>POST /client/v4/accounts/:accountId/workers/scripts/:scriptName/assets-upload-session
+    User<<->>Workers API: Upload files<br/>POST /client/v4/accounts/:accountId/workers/assets/upload?base64=true
+    User<<->>Workers API: Upload script version<br/>PUT /client/v4/accounts/:accountId/workers/scripts/:scriptName
+```
+
+The asset upload flow can be distilled into three distinct phases: registration of a manifest, uploading of the assets, and deployment of the Worker.
```
Comment (Oxyjun) [inline src/content/docs/workers/static-assets/direct-upload.mdx]:
```suggestion
The asset upload flow can be distilled into three distinct phases:

1. Registration of a manifest
2. Upload of the assets
3. Deployment of the Worker
```

File: src/content/docs/workers/static-assets/direct-upload.mdx [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+```bash title="Example manifest upload request"
+curl -X POST https://api.cloudflare.com/client/v4/accounts/:accountId/workers/scripts/:scriptName/assets-upload-session \
+ --header 'content-type: application/json' \
+ --header 'Authorization: Bearer API_TOKEN' \
+ --data '{ "manifest": { "/filea.html": { "hash": "08f1dfda4574284ab3c21666d1", "size": 12 }, "/fileb.html": { "hash": "4f1c1af44620d531446ceef93f", "size": 23 } } }'
+```
+
+The resulting response will contain a JWT, which provides authentication during file upload. The JWT is valid for one hour.
+
+In addition to the JWT, the response instructs users how to most optimally batch upload their files. These instructions are encoded in the "buckets" field. Each array in buckets contains a list of file hashes which should be uploaded together.
```
Comment (Oxyjun) [inline src/content/docs/workers/static-assets/direct-upload.mdx]:
```suggestion
In addition to the JWT, the response instructs users how to optimally batch upload their files. These instructions are encoded in the `buckets` field. Each array in buckets contains a list of file hashes which should be uploaded together.
```

File: src/content/docs/workers/static-assets/direct-upload.mdx [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+
+Users can upload assets by POSTing to https://api.cloudflare.com/client/v4/accounts/:accountID/workers/assets/upload?base64=true. This endpoint requires files be uploaded using multipart/form data. The contents of each file must be base64 encoded, and the `base64` query param must be set to `true`.
+
+The authorization header must be provided as a bearer token, using the JWT (upload token) from the aforementioned manifest upload call.
+
+Once every file in the manifest has been uploaded, a status code of 201 will be returned, with the `jwt` field present, and containing a final "completion" token which can be used in a deployment of a Worker with assets. This token is valid for 1 hour.
+
+## `Create/Deploy New Version`
+
+[Script](https://developers.cloudflare.com/api/operations/worker-script-upload-worker-module) and [version](https://developers.cloudflare.com/api/operations/worker-versions-upload-version) upload endpoints require specifying a metadata part in the form data. If this is a new Worker, we can provide the completion token from the previous (upload assets) step. This token will be validated by the backend to ensure that all assets have been uploaded.
```
Comment (Oxyjun) [inline src/content/docs/workers/static-assets/direct-upload.mdx]:
```suggestion
[Script](/api/operations/worker-script-upload-worker-module) and [version](/api/operations/worker-versions-upload-version) upload endpoints require specifying a metadata part in the form data. If this is a new Worker, we can provide the completion token from the previous (upload assets) step. This token will be validated by the backend to ensure that all assets have been uploaded.
```

File: src/content/docs/workers/static-assets/direct-upload.mdx [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+
+```bash title="Example Worker Metadata Specifying keep_assets"
+{
+  "main_module": "main.js",
+  "keep_assets": true,
+  "compatibility_date": "2021-09-14"
+}
+```
+
+Asset [routing configuration](/workers/static-assets/routing/#routing-configuration) can be provided in the assets object, such as html_handling and not_found_handling.
```
Comment (Oxyjun) [inline src/content/docs/workers/static-assets/direct-upload.mdx]:
```suggestion
Asset [routing configuration](/workers/static-assets/routing/#routing-configuration) can be provided in the assets object, such as `html_handling` and `not_found_handling`.
```

**Review by Oxyjun — APPROVED** [review]

File: src/content/docs/workers/static-assets/direct-upload.mdx (hunk 2 above)
Comment (WillTaylorDev) [reply src/content/docs/workers/static-assets/direct-upload.mdx]: Merged, and url updated

File: src/content/docs/workers/static-assets/direct-upload.mdx [hunk 8]
```diff
@@ -0,0 +1,365 @@
...
+    User<<->>Workers API: Upload script version<br/>PUT /client/v4/accounts/:accountId/workers/scripts/:scriptName
+```
+
+The asset upload flow can be distilled into three distinct phases:
+
+1. Registration of a manifest
+2. Upload of the assets
+3. Deployment of the Worker
+
+## `Upload manifest`
```
Comment (GregBrimble) [inline src/content/docs/workers/static-assets/direct-upload.mdx]:
```suggestion
## Upload manifest
```

File: src/content/docs/workers/static-assets/direct-upload.mdx [hunk 9]
```diff
@@ -0,0 +1,365 @@
...
+      ]
+    ]
+  },
+  "success": true,
+  "errors": null,
+  "messages": null
+}
+```
+
+### `Limitations`
```
Comment (GregBrimble) [inline src/content/docs/workers/static-assets/direct-upload.mdx]:
```suggestion
### Limitations
```

File: src/content/docs/workers/static-assets/direct-upload.mdx [hunk 10]
```diff
@@ -0,0 +1,365 @@
...
+
+- Each file must be under 25 MiB
+- The overall manifest must not contain more than 20,000 file entries
+
+## `Upload Static Assets`
```
Comment (GregBrimble) [inline src/content/docs/workers/static-assets/direct-upload.mdx]:
```suggestion
## Upload Static Assets
```

File: src/content/docs/workers/static-assets/direct-upload.mdx [hunk 11]
```diff
@@ -0,0 +1,365 @@
...
+
+The [file upload API](/api/operations/worker-assets-upload) requires files be uploaded using multipart/form data. The contents of each file must be base64 encoded, and the `base64` query param must be set to `true`.
+
+The authorization header must be provided as a bearer token, using the JWT (upload token) from the aforementioned manifest upload call.
+
+Once every file in the manifest has been uploaded, a status code of 201 will be returned, with the `jwt` field present, and containing a final "completion" token which can be used in a deployment of a Worker with assets. This token is valid for 1 hour.
+
+## `Create/Deploy New Version`
```
Comment (GregBrimble) [inline src/content/docs/workers/static-assets/direct-upload.mdx]:
```suggestion
## Create/Deploy New Version
```

File: src/content/docs/workers/static-assets/direct-upload.mdx [hunk 12]
```diff
@@ -0,0 +1,365 @@
...
+      "name": "ASSETS",
+      "type": "assets"
+    }
+    ...
+  ]
+  "compatibility_date": "2021-09-14"
+}
+```
+
+## `Programmatic Example`
```
Comment (GregBrimble) [inline src/content/docs/workers/static-assets/direct-upload.mdx]:
```suggestion
## Programmatic Example
```

File: src/content/docs/workers/static-assets/direct-upload.mdx [hunk 13]
```diff
@@ -0,0 +1,365 @@
...
+	FileTree,
+	InlineBadge,
+	Render,
+	TabItem,
+	Tabs,
+} from "~/components";
+
+:::note
+
+Directly uploading assets via Worker APIs is an advanced approach that most users will not need. Instead, we encourage users to interact with these APIs through [Wrangler](/workers/static-assets/get-started/#1-create-a-new-worker-project-using-the-cli).
```
Comment (GregBrimble) [inline src/content/docs/workers/static-assets/direct-upload.mdx]:
```suggestion
Directly uploading assets via APIs is an advanced approach that most users will not need. Instead, we encourage users to deploy your Worker with [Wrangler](/workers/static-assets/get-started/#1-create-a-new-worker-project-using-the-cli).
```

File: src/content/docs/workers/static-assets/direct-upload.mdx [hunk 14]
```diff
@@ -0,0 +1,365 @@
...
+
+:::
+
+The Workers API empowers users to upload and serve static assets. Users can also fetch assets through an optional [assets binding](/workers/static-assets/binding/). This guide will describe the process for attaching assets to your Worker directly through the Workers API.
```
Comment (GregBrimble) [inline src/content/docs/workers/static-assets/direct-upload.mdx]:
```suggestion
Our API empowers users to upload and include static assets as part of a Worker. These static assets can be served for free, and additionally, users can also fetch assets through an optional [assets binding](/workers/static-assets/binding/) to power more advanced applications. This guide will describe the process for attaching assets to your Worker directly with the API.
```

File: src/content/docs/workers/static-assets/direct-upload.mdx [hunk 15]
```diff
@@ -0,0 +1,365 @@
...
+
+## `Upload manifest`
+
+The asset manifest is a ledger which keeps track of files we want to use in our Worker. This manifest is used to track assets associated with each Worker version, and eliminate the need to upload unchanged files prior to a new upload.
+
+The [manifest upload request](/api/operations/worker-script-update-create-assets-upload-session) describes each file which we intend to upload. Each file is its own key representing the file path and name, and is an object which contains metadata about the file.
+
+`hash` represents the first 32 characters of the file's hash, while `size` is the size (in bytes) of the file.
+
+```bash title="Example manifest upload request"
```
Comment (GregBrimble) [inline src/content/docs/workers/static-assets/direct-upload.mdx]: Please format these with the style-guide-approved placeholder formats in the headers and URI.

https://developers.cloudflare.com/style-guide/api-content-strategy/guidelines-for-curl-commands/#request-guidelines

File: src/content/docs/workers/static-assets/direct-upload.mdx [hunk 16]
```diff
@@ -0,0 +1,365 @@
...
+curl -X POST https://api.cloudflare.com/client/v4/accounts/:accountId/workers/scripts/:scriptName/assets-upload-session \
+ --header 'content-type: application/json' \
+ --header 'Authorization: Bearer API_TOKEN' \
+ --data '{ "manifest": { "/filea.html": { "hash": "08f1dfda4574284ab3c21666d1", "size": 12 }, "/fileb.html": { "hash": "4f1c1af44620d531446ceef93f", "size": 23 } } }'
+```
+
+The resulting response will contain a JWT, which provides authentication during file upload. The JWT is valid for one hour.
+
+In addition to the JWT, the response instructs users how to optimally batch upload their files. These instructions are encoded in the `buckets` field. Each array in buckets contains a list of file hashes which should be uploaded together.
```
Comment (GregBrimble) [inline src/content/docs/workers/static-assets/direct-upload.mdx]:
```suggestion
In addition to the JWT, the response instructs users how to optimally batch upload their files. These instructions are encoded in the `buckets` field. Each array in `buckets` contains a list of file hashes which should be uploaded together.
```

File: src/content/docs/workers/static-assets/direct-upload.mdx [hunk 17]
```diff
@@ -0,0 +1,365 @@
...
+
+```bash title="Example manifest upload response"
+{
+  "result": {
+    "jwt": "<upload_token>",
+    "buckets": [
+      [
+        "08f1dfda4574284ab3c21666d1",
+				"4f1c1af44620d531446ceef93f"
```
Comment (GregBrimble) [inline src/content/docs/workers/static-assets/direct-upload.mdx]: Worth maybe rendering this as two arrays? Maybe include a third file in your example so it goes 2 and 1? Just might make it a little clearer for people that this is an array of arrays, ultimately containing all files.

File: src/content/docs/workers/static-assets/direct-upload.mdx (hunk 16 above)
Comment (GregBrimble) [inline src/content/docs/workers/static-assets/direct-upload.mdx]: Worth including a note saying that some hashes may not be returned? This is a best-effort system and will happen when we know we've seen an asset with that hash recently.

File: src/content/docs/workers/static-assets/direct-upload.mdx [hunk 18]
```diff
@@ -0,0 +1,365 @@
...
+2. Upload of the assets
+3. Deployment of the Worker
+
+## `Upload manifest`
+
+The asset manifest is a ledger which keeps track of files we want to use in our Worker. This manifest is used to track assets associated with each Worker version, and eliminate the need to upload unchanged files prior to a new upload.
+
+The [manifest upload request](/api/operations/worker-script-update-create-assets-upload-session) describes each file which we intend to upload. Each file is its own key representing the file path and name, and is an object which contains metadata about the file.
+
+`hash` represents the first 32 characters of the file's hash, while `size` is the size (in bytes) of the file.
```
Comment (GregBrimble) [inline src/content/docs/workers/static-assets/direct-upload.mdx]:
```suggestion
`hash` represents a 32 hexadecimal character hash of the file, while `size` is the size (in bytes) of the file.
```

File: src/content/docs/workers/static-assets/direct-upload.mdx [hunk 19]
```diff
@@ -0,0 +1,365 @@
...
+```
+
+### `Limitations`
+
+- Each file must be under 25 MiB
+- The overall manifest must not contain more than 20,000 file entries
+
+## `Upload Static Assets`
+
+The [file upload API](/api/operations/worker-assets-upload) requires files be uploaded using multipart/form data. The contents of each file must be base64 encoded, and the `base64` query param must be set to `true`.
```
Comment (GregBrimble) [inline src/content/docs/workers/static-assets/direct-upload.mdx]:
```suggestion
The [file upload API](/api/operations/worker-assets-upload) requires files be uploaded using `multipart/form-data`. The contents of each file must be base64 encoded, and the `base64` query parameter in the URL must be set to `true`.
```

File: src/content/docs/workers/static-assets/direct-upload.mdx [hunk 20]
```diff
@@ -0,0 +1,365 @@
...
+
+The authorization header must be provided as a bearer token, using the JWT (upload token) from the aforementioned manifest upload call.
```
Comment (GregBrimble) [inline src/content/docs/workers/static-assets/direct-upload.mdx]:
```suggestion
The `Authorization` header must be provided as a bearer token, using the JWT (upload token) from the aforementioned manifest upload call.
```

File: src/content/docs/workers/static-assets/direct-upload.mdx [hunk 21]
```diff
@@ -0,0 +1,365 @@
...
+
+Once every file in the manifest has been uploaded, a status code of 201 will be returned, with the `jwt` field present, and containing a final "completion" token which can be used in a deployment of a Worker with assets. This token is valid for 1 hour.
```
Comment (GregBrimble) [inline src/content/docs/workers/static-assets/direct-upload.mdx]:
```suggestion
Once every file in the manifest has been uploaded, a status code of 201 will be returned, with the `jwt` field present. This JWT is a final "completion" token which can be used to create a deployment of a Worker with this set of assets. This completion token is valid for 1 hour.
```

File: src/content/docs/workers/static-assets/direct-upload.mdx [hunk 22]
```diff
@@ -0,0 +1,365 @@
...
+
+## `Create/Deploy New Version`
+
+[Script](/api/operations/worker-script-upload-worker-module) and [version](/api/operations/worker-versions-upload-version) upload endpoints require specifying a metadata part in the form data. If this is a new Worker, we can provide the completion token from the previous (upload assets) step. This token will be validated by the backend to ensure that all assets have been uploaded.
```
Comment (GregBrimble) [inline src/content/docs/workers/static-assets/direct-upload.mdx]:
```suggestion
[Script](/api/operations/worker-script-upload-worker-module) and [version](/api/operations/worker-versions-upload-version) upload endpoints require specifying a metadata part in the form data. Here, we can provide the completion token from the previous (upload assets) step.
```

File: src/content/docs/workers/static-assets/direct-upload.mdx [hunk 23]
```diff
@@ -0,0 +1,365 @@
...
+{
+  "main_module": "main.js",
+  "assets": {
+    "jwt": "<completion_token>"
+  },
+  "compatibility_date": "2021-09-14"
+}
+```
+
+If this is a worker which already exists with assets, and you wish to just re-use the existing set of assets, we do not have to specify the completion token again. Instead, we can provide the `keep_assets` metadata configuration.
```
Comment (GregBrimble) [inline src/content/docs/workers/static-assets/direct-upload.mdx]:
```suggestion
If this is a Worker which already has assets, and you wish to just re-use the existing set of assets, we do not have to specify the completion token again. Instead, we can pass the boolean `keep_assets` option.
```

File: src/content/docs/workers/static-assets/direct-upload.mdx [hunk 24]
```diff
@@ -0,0 +1,365 @@
...
+
+```bash title="Example Worker Metadata Specifying keep_assets"
+{
+  "main_module": "main.js",
+  "keep_assets": true,
+  "compatibility_date": "2021-09-14"
+}
+```
+
+Asset [routing configuration](/workers/static-assets/routing/#routing-configuration) can be provided in the assets object, such as `html_handling` and `not_found_handling`.
```
Comment (GregBrimble) [inline src/content/docs/workers/static-assets/direct-upload.mdx]:
```suggestion
Asset [routing configuration](/workers/static-assets/routing/#routing-configuration) can be provided in the `assets` object, such as `html_handling` and `not_found_handling`.
```

File: src/content/docs/workers/static-assets/direct-upload.mdx [hunk 25]
```diff
@@ -0,0 +1,365 @@
...
+    "jwt": "<completion_token>",
+    "config" {
+      "html_handling": "auto-trailing-slash"
+    }
+  },
+  "compatibility_date": "2021-09-14"
+}
+```
+
+Optionally, an asset binding can be provided if you wish to fetch and serve assets from within your Worker.
```
Comment (GregBrimble) [inline src/content/docs/workers/static-assets/direct-upload.mdx]:
```suggestion
Optionally, an assets binding can be provided if you wish to fetch and serve assets from within your Worker code.
```

File: src/content/docs/workers/static-assets/direct-upload.mdx [hunk 26]
```diff
@@ -0,0 +1,365 @@
...
+<Tabs> <TabItem label="TypeScript" icon="seti:typescript">
+
+```ts
+import * as fs from "fs";
+import * as path from "path";
+import * as crypto from "crypto";
+import { FormData, fetch } from "undici";
+
+const accountId: string = ""; // Replace with your actual account ID
+const authToken: string = ""; // Replace with your actual API token
```
Comment (GregBrimble) [inline src/content/docs/workers/static-assets/direct-upload.mdx]: Can we swap this out for a `process.env.CLOUDFLARE_API_TOKEN` thing throughout this script? I know this is just a throwaway example, but it's important to have security in-mind for people that just blindly copy-paste.

File: src/content/docs/workers/static-assets/direct-upload.mdx [hunk 27]
```diff
@@ -0,0 +1,365 @@
...
+	result: {
+		jwt: string;
+		buckets: string[][];
+	};
+	success: boolean;
+	errors: any;
+	messages: any;
+}
+
+// Function to calculate the SHA-256 hash of a file and truncate to 32 bits
```
Comment (GregBrimble) [inline src/content/docs/workers/static-assets/direct-upload.mdx]:
```suggestion
// Function to calculate the SHA-256 hash of a file and truncate to 32 characters
```

File: src/content/docs/workers/static-assets/direct-upload.mdx [hunk 28]
```diff
@@ -0,0 +1,365 @@
...
+			},
+		);
+
+		const data = (await response.json()) as UploadResponse;
+		if (data && data.result.jwt) {
+			return { completionToken: data.result.jwt };
+		}
+	});
+
+	throw new Error("Should have received completion token");
```
Comment (GregBrimble) [inline src/content/docs/workers/static-assets/direct-upload.mdx:290]: Isn't that just going to always fire?

File: src/content/docs/workers/static-assets/direct-upload.mdx [hunk 29]
```diff
@@ -0,0 +1,365 @@
...
+			assets: {
+				jwt: completionToken, // Provide the completion token from file uploads
+			},
+			bindings: [{ name: "ASSETS", type: "assets" }], // Optional assets binding to fetch from user worker
+		}),
+	);
+
+	// Configure (optional) user worker
+	form.append(
+		"@index.js",
```
Comment (GregBrimble) [inline src/content/docs/workers/static-assets/direct-upload.mdx]: Unusual to put a `@` here. Is that needed? Leftovers from the cURL syntax of specifying a local file?

File: src/content/docs/workers/static-assets/direct-upload.mdx [hunk 30]
```diff
@@ -0,0 +1,365 @@
...
+	});
+
+	const response = await fetch(
+		`https://api.cloudflare.com/client/v4/accounts/${accountId}/workers/scripts/${scriptName}/assets-upload-session`,
+		{
+			method: "POST",
+			headers: {
+				Authorization: `Bearer ${authToken}`,
+				"Content-Type": "application/json",
+				"Content-Length": Buffer.byteLength(requestBody).toString(),
```
Comment (GregBrimble) [inline src/content/docs/workers/static-assets/direct-upload.mdx]: Wouldn't bother with this. Fetch API will attach a Content-Length header for you, I think: https://cloudflareworkers.com/#16ff10a7c51d218f288a821ccff47a81:https://tutorial.cloudflareworkers.com

```suggestion
```

File: src/content/docs/workers/static-assets/direct-upload.mdx [hunk 31]
```diff
@@ -0,0 +1,365 @@
...
+		uploadToken: jwt,
+		buckets: data.result.buckets,
+		fileMetadata,
+	};
+}
+
+// Begin the upload session by uploading a new manifest
+const { uploadToken, buckets, fileMetadata } = await startUploadSession();
+
+// If all files are already uploaded, a completion token will be immediately returned. Otherwise,
```
Comment (GregBrimble) [inline src/content/docs/workers/static-assets/direct-upload.mdx:372]: Ah, forgot about that. That's maybe something we want to mention in the written documentation above too.

File: src/content/docs/workers/static-assets/direct-upload.mdx [hunk 32]
```diff
@@ -0,0 +1,365 @@
...
+// we should upload the missing files
+let completionToken = uploadToken;
+if (buckets.length > 0) {
+	completionToken = await uploadFilesBatch(uploadToken, buckets, fileMetadata);
+}
+
+// Once we have uploaded all of our files, we can upload a new script, and assets, with completion token
+scriptUpload(completionToken);
```
Comment (GregBrimble) [inline src/content/docs/workers/static-assets/direct-upload.mdx]:
```suggestion
await scriptUpload(completionToken);
```

File: src/content/docs/workers/static-assets/direct-upload.mdx (hunk 28 above)
Comment (WillTaylorDev) [reply src/content/docs/workers/static-assets/direct-upload.mdx:290]: This function takes in ALL buckets, and iterates over them until it gets a completion token; if we get to this point, we've done all the file uploads, but no token came back from AUS. Think this is right!

File: src/content/docs/workers/static-assets/direct-upload.mdx (hunk 29 above)
Comment (WillTaylorDev) [reply src/content/docs/workers/static-assets/direct-upload.mdx]: Ah, yes indeed-y. Thank you!

**Comment by WillTaylorDev** [comment]
Thanks @GregBrimble, feedback all addressed!

File: src/content/docs/workers/static-assets/direct-upload.mdx (hunk 28 above)
Comment (GregBrimble) [reply src/content/docs/workers/static-assets/direct-upload.mdx:290]: Does the `return` on L286 break out of the `forEach`? I don't think so..?

https://workers.cloudflare.com/playground#LYVwNgLglgDghgJwgegGYHsHALQBM4RwDcABAEbogB2+CAngLzbPYDqApmQNJQQBimYACFKNRHQASAeQAaEgB4AZAJIA1GQAUA1gHchG4AHYAXCxYduvAVhHVak2QpXrteg4YCwAKADC6KhDsAdgAIlAAzjDo4bxQ-sYkGFh4BMQkVHDA7AwARPiEAHQAVuE5pKhQYEGZ2XmpxaXefgFBENgAKnQw7AlwMDBgUADGBHFUyEVwAG5w4UMIsBAA1MDouODs3t7s8lFIJLjsqHDgECQA3l4AkBlZCTkAovKZA+wkAIIayjkANNeEAHNwgkANo5HYvKq-Eg5PpQaE5BDsSL+cLsHIAXT+AF8iF5vGZmBYePxBLYxPRpHIlGpNLp9EYmv5AsEwiiYtB4olBClCKRbrUoDQdg0yolKtUsrkhYd5KKmS1gp1ur1+oMRpzxpMZnMFjBlqt1lUtl4oMA9mcdOwwEN0FlEgg7TCrTa7ewCgALCDAMBlLxeZAAKkDXhIgZIAAEIF12IdUBcpGQiuwhhBsSQHlQpqHA8h-TsLQcjidIBdrkGQ1cw5H4AhMhcAErsACOIGRaZISNb7eu1Yjtfr50zU3TQWzVfD-cQg6eKZAmuagXkHdT8l7k6REBACCo4QuGkdwAi7AAPE2Ubv2AA+bHrvNXWZ0KhDRLsCBDD0ACi7bfCEB+JBjgBq4AJRllcVy2ruZyIAgJAMCQIIAORwEhAFIWQaEkEhQxIRi-oQbBBRJA8cAfp+n5QGBDBXuBVYkFA8aUfBCEYUhYGXBBVybtuVDpOwOgkOeUSXp+SEelA7HXPRt4QdiIEEVcEAeo6glUAJGYII6CCfjkZBwFQAI5ApVzYjieIEoSxJWGSoj2FSTi0q4DKeL4zKtKEERRByYwJEkOD5GkAq5C6tpZJ63q+uUErBTkoVuhFPo5AqLJtMqPQkEuKADHAQomieACEIRSD47QAJoaA8JBej6V5eCeNVgCQYAGQCuRBDkdVXA17BwLgXXdVkhAkB+iBohAuQAKrtHw2AABw5CQyADSeQ1wNVEAQDA2AtiAUBTLkMjYJN7zYH45qjGQUIje5AS5MoDwMLGALoktK1rekNS5FMUACRai1QalIVQLgykMIcP1DOw2A6CDykAUKsRwGA2BzMj2QAIwFAADIty3SSe0AQFUV4+GAlC4KgLVIiQrCYFo7AIHuGgtXQAKOnYJ7IETJME4MVBaNJ3HWrkf50FU4Qeuwb45ELKlHLkXpbcCyDIDlbMczQBQQzAjq4AUQzkyAlPU+6EPIPFWSo9GVQG+EpTSfj3XIFLfV1dcJ4ULgdArbg+0jS19u5FTiDIp1QsnmaAIkOECBDIrm0wCraus+zdna+wUy62sBtGybocZ1MaCm+EBQwIZeMrcgftTO9uVUAN3U1wHszhCF9OM+E2DkwC6CdVzNeNyeMCN1cHCulkBUkJNaKZRJzOp5ruCZegmXtgcCxTG8610wgDMIMBSIENvRarCvQtXJLockP4-PsABBnLzoUt8Uf3tFgM6B0AcUBIqmYDfwgKvZSbxnpvDIN-ca0BDIkBADAEgcAL5kwpiHGmZFbTUAgAUCOash4egxleVgHoCAMT3DvDuCAAD8XN8FDxHhfd4JBkHG1QW8Xe+9SEkAAFLTDgAAZXmIsG6hwSB0EoCQHQCxAhzxIcQmgEtRGUDglaMgF8OTsCQnuCQ7R2gaEynWVAFQXx+z-sTb+QomF51YZowCuAXrNXQBqMYZCl5z3YBfHQmAwC4AfmAcmsMYFiJACvBxGo3i2hEYbaIbwgGKO3IBGgsC0RMwYnxTAhw4JAIvkiC8y9YkgOACQVYNNWzDC0AAnB9CuINQIe0egDEzhBKZtaVANCCERyqVxKQfEmnNSOGcCICCY6QjYRQmRAy9wIGoFQIU0d-Bz0GeogoJAyriJGFQC+sZeANOavtbeiS0QxKlp2ZEpw9xSyRMsh4fszggJIPAZSCCKBbxCcgOE58uKX2ltVdAgk7kQnNFUWmYy5G4AUX7QxjNWidkoIEUuKy1kGQQWAcIq8tkQAvnciJMTV6HIkcQiANjvIxCuvs5eZAFiBKUfxZcDFDjrViYMVA7pKlV2APXLqXMvY+3qi7SKdVLJmGsqSGwdlxAORpC4ek7gUoeTZN5WIXJ-K8iCl9HIa1ApigqFUWKV10BkGSm5RUaUYyqgGMMUY-gJioqoCac4OQwoPKgGQSovA6AAH18jomMDkAATNjX1AAWbA2MADMIbDC-AdXaJ1LrBjRndSHIEORjBgioGsdgJR3WOoIJiH46r67usNBsFNOQZQihKDkbEgrzCcBJNYYQYrKSOElXSNwRhmDeCAA

File: src/content/docs/workers/static-assets/direct-upload.mdx [hunk 33]
```diff
@@ -0,0 +1,383 @@
...
+
+:::
+
+Our API empowers users to upload and include static assets as part of a Worker. These static assets can be served for free, and additionally, users can also fetch assets through an optional [assets binding](/workers/static-assets/binding/) to power more advanced applications. This guide will describe the process for attaching assets to your Worker directly with the API.
+
+```mermaid
+sequenceDiagram
+    participant User
+    participant Workers API
+    User<<->>Workers API: Submit manifest<br/>POST /client/v4/accounts/:accountId/workers/scripts/:scriptName/assets-upload-session
```
Comment (GregBrimble) [inline src/content/docs/workers/static-assets/direct-upload.mdx:32]: Use the same placeholder syntax as the cURL? Idk. whatever :)

**Review by GregBrimble — APPROVED** [review]

File: src/content/docs/workers/static-assets/direct-upload.mdx (hunk 28 above)
Comment (WillTaylorDev) [reply src/content/docs/workers/static-assets/direct-upload.mdx:290]: Thanks Greg for the help! We got it 💪

**Pull request merged**